proposed/2021/TargetFrameworkAndPlatformsFilters.md

```diff
@@ -0,0 +1,205 @@
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+When a developer is browsing for a NuGet package to include in their project, they can filter packages based on generations of .NET including the target framework & supported platforms that package targets.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Finding a package that is compatible with your project is a daily challenge for the average developer in the .NET ecosystem. Over 21% of developers fail to install a package today because of an incompatibility; it's one of NuGet's top errors known as `NU1202`. On top of it being one of the biggest error codes we see, it is also [the number #1 request on the `NuGetGallery` repository](https://github.com/NuGet/NuGetGallery/issues/3098) for good reason. It should be straight-forward for a developer to separate the generation of .NET they are currently targeting and find packages that are supported on the platforms that the respective generation supports.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+When browsing for packages on NuGet.org, a developer will be able to filter packages based on compatibility filters. These filters will include:
+
+1. The **generation of .NET** such as .NET Framework, .NET Core, .NET (5, 6, and later), and Any.
+2. The **supported platforms** for the generation such as .NET Standard, Windows, macOS, Linux, Android, iOS, and more.
+3. The **version number** of the generation being used such as .NET 5, .NET Core 3.1, .NET Framework 4.8, and more.
+
+This proposal only accounts for the **majority of packages on NuGet.org**. There are many unique TFMs in the .NET ecosystem, but for the sake of creating a concept that will scale with the future of .NET, many unsupported & low usage TFMs are not considered. Thus this proposal only focuses on the [most common target frameworks](https://docs.microsoft.com/dotnet/standard/frameworks#latest-versions) & omits considering TFMs with fewer than 5,000 packages.
+
+#### Generations of .NET
+
+There are many generations of .NET. For the sake of this proposal, we will focus on defining the three major generations:
+
+- .NET Framework:
+  - `net, net11, net20, net35, net40, net403, net45, net451, net452, net46, net461, net462, net463, net47, net471, net472, net48`
+- .NET Core:
+  - `netcoreapp1.0, netcoreapp1.1, netcoreapp2.0, netcoreapp2.1, netcoreapp2.2, netcoreapp3.0, netcoreapp3.1`
+- .NET:
+  - `net5.0, net6.0`
+- Any:
+  - .NET Framework, .NET Core, .NET
+
+![](meta/resources/TargetFrameworkAndPlatformFilters/GenerationsOfNET.png)
+
+#### Supported Platforms
```

Comment:
@advay26 and I talked a bit about maybe phasing out platforms and frameworks. Today, the proposed UI for the nuget.org package details page only refers to frameworks. There is no platform notion for display yet. We could consider implementing first the framework filtering to align with the upcoming package details page changes and only add the platform filtering when it's materialized in a well understood way on the details page.

As an aside, the platform concept is a bit more opaque to me (perhaps due to my high exposure to NuGet-only concepts) so I'm wondering if we actually have consensus in the .NET ecosystem on what "platform" means.

Reply:
This spec leapfrogs a bit of package details as it was written before it. I believe we can revisit package details with inclusion of platforms or have a feature flag for platforms until package details has parity.

**I've added that as an opportunity at the bottom of the spec just now.**

We definitely want a consistent experience & I'll visit that again shortly.

How .NET defines a platform is the following:

```
platform for operating systems and chipsets
(Windows, macOS, Linux) with distros for Linux
flavors. Unless prefixed by “developer” which
can mean a programming platform
```
from https://github.com/dotnet/brand/blob/main/dotnet-brand-guidelines.pdf

![image](https://user-images.githubusercontent.com/1288848/145269528-1d982fa5-1c14-4359-9366-64e39a5f1443.png)

Really what we're trying to do here is break up TFM strings from being opaque to having individual meaning in parts. Such as this scenario from .NET 5:

https://github.com/dotnet/designs/blob/main/accepted/2020/net5/net5.md#valid-platforms

> So with a TargetFramework of net5.0-android, NuGet should parse out the platform as Android. The core .NET SDK will then set the TargetPlatformIdentifier to android. The Android workload should then normalize the casing of the TargetPlatformIdentifier property to Android. Similarly, the iOS Workload and Windows targets should normalize the TargetPlatformIdentifier to iOS and Windows.

For filtering purposes, we want to show packages that have support for that platform with platform-specific assets and/or .NET Standard(portable) assets if selected.